dtls.c

```diff
@@ -1539,6 +1510,98 @@ dtls_prepare_record(dtls_peer_t *peer, dtls_security_parameters_t *security,
+			     dtls_ephemeral_peer_t *ephemeral_peer,
+			     dtls_alert_level_t level,
+			     dtls_alert_t description)
+{
+  uint8 buf[DTLS_RH_LENGTH + DTLS_ALERT_LENGTH];
+  uint8 *p = dtls_set_record_header(DTLS_CT_HANDSHAKE, 0, &(ephemeral_peer->rseq), buf);
+
+  /* fix length of fragment in sendbuf */
+  dtls_int_to_uint16(buf + 11, DTLS_ALERT_LENGTH);
+
+  /* Alert */
+  dtls_int_to_uint8(p, level);
+  dtls_int_to_uint8(p + 1, description);
+
+  dtls_debug("send alert - protocol version  packet\n");
+
+  dtls_debug_hexdump("send header", buf, DTLS_RH_LENGTH);
+  dtls_debug_hexdump("send unencrypted alert", p, DTLS_ALERT_LENGTH);
+
+  int res = CALL(ctx, write, ephemeral_peer->session, buf, sizeof(buf));
+
+  /* Guess number of bytes application data actually sent:
+   * dtls_prepare_record() tells us in len the number of bytes to
+   * send, res will contain the bytes actually sent. */
+  return res <= 0 ? res : (int) sizeof(buf);
+}
+
+/**
+ * Send HelloVerifyRequest to initial challenge a peer in a stateless fashion.
+ * A HelloVerifyRequest is sent to the peer (using the write callback function
+ * registered with \p ctx). The return value is \c -1 on error, \c 1 when
+ * the HelloVerifyRequest is sent.
+ *
+ * \param ctx              The DTLS context.
+ * \param ephemeral_peer   The ephemeral remote party we are talking to.
+ * \param data             The received datagram.
+ * \param data_length      Length of \p msg.
+ * \return \c 1 if msg is sent, or \c -1 otherwise.
+ */
+static int
+dtls_0_send_hello_verify_request(dtls_context_t *ctx,
+			     dtls_ephemeral_peer_t *ephemeral_peer,
+			     uint8 *data, size_t data_length)
+{
+  int res;
+  uint8 buf[DTLS_RH_LENGTH + DTLS_HS_LENGTH + data_length];
+  uint8 *p = dtls_set_record_header(DTLS_CT_HANDSHAKE, 0, &(ephemeral_peer->rseq), buf);
+
+  /* Signal DTLS version 1.0 in the record layer of ClientHello and
+   * HelloVerifyRequest handshake messages according to Section 4.2.1
+   * of RFC 6347.
+   *
+   * This does not apply to a renegotation ClientHello
+   */
+  dtls_int_to_uint16(buf + 1, DTLS10_VERSION);
+
+  /* fix length of fragment in sendbuf */
+  dtls_int_to_uint16(buf + 11, DTLS_HS_LENGTH + data_length);
+
+  p = dtls_set_handshake_header(DTLS_HT_HELLO_VERIFY_REQUEST, &(ephemeral_peer->mseq), data_length, 0,
+			    data_length, p);
+
+  memcpy(p, data, data_length);
+
+  dtls_debug("send hello_verify_request packet\n");
+
+  dtls_debug_hexdump("send header", buf, DTLS_RH_LENGTH);
+  dtls_debug_hexdump("send unencrypted handshake header", buf + DTLS_RH_LENGTH, DTLS_HS_LENGTH);
+  dtls_debug_hexdump("send unencrypted cookie", data, data_length);
+
+  res = CALL(ctx, write, ephemeral_peer->session, buf, sizeof(buf));
+
+  /* Guess number of bytes application data actually sent:
+   * dtls_prepare_record() tells us in len the number of bytes to
+   * send, res will contain the bytes actually sent. */
+  return res <= 0 ? res : (int) sizeof(buf);
```

Comment:
Here, the same holds as for the doxygen comment and return statement of `dtls_0_send_alert()`.

Reply:
OK